File: progs/javascript/wait-n.js
const id = async (n) => {
  return n;
};

const waitN = async (futs) => {
  return await Promise.all(futs);
};

const main = async () => {
  const timer = process.hrtime.bigint();
  const futs = Array.from({ length: 1001 }, (_, i) => id(i));
  const ns = await waitN(futs);
  const res = ns.reduce((acc, val) => acc + val, 0);
  console.assert(res === 500500, 'Result does not match expected value');
  const elapsed = Number(process.hrtime.bigint() - timer) / 1000;
  console.log(`done ${elapsed}μs`);
};

main().catch(console.error);

File: progs/python/misc/wait-n.py
import asyncio
import time

async def id(n):
    return n

async def wait_n(futs):
    return await asyncio.gather(*futs)

async def main():
    timer = time.perf_counter()
    futs = (id(i) for i in range(1001))
    ns = await wait_n(futs)
    res = sum(ns)
    assert res == 500500
    print(f"done {int((time.perf_counter() - timer) * 1_000_000)}μs")

if __name__ == "__main__":
    asyncio.run(main())
